lib/ruby_lsp/requests/document_highlight.rb

```diff
@@ -3,86 +3,739 @@
+        end
+
+        sig { params(node: YARP::LocalVariableOperatorWriteNode).void }
+        def on_local_variable_operator_write(node)
+          add_highlight(Highlight.new(kind: WRITE, location: node.name_loc)) if node.name == name
+        end
+
+        sig { params(node: YARP::OptionalParameterNode).void }
+        def on_optional_parameter(node)
+          add_highlight(Highlight.new(kind: WRITE, location: node.name_loc)) if node.name == name
+        end
+
+        sig { params(node: YARP::RequiredParameterNode).void }
+        def on_required_parameter(node)
+          add_highlight(Highlight.new(kind: WRITE, location: node.location)) if node.name == name
+        end
+
+        sig { params(node: YARP::RestParameterNode).void }
+        def on_rest_parameter(node)
+          add_highlight(Highlight.new(kind: WRITE, location: T.must(node.name_loc))) if node.name == name
+        end
       end
 
-      sig { params(node: T.nilable(YARP::Node)).void }
-      def on_node(node)
-        return if node.nil?
+      # ![Document highlight demo](../../document_highlight.gif)
+      #
+      # The [document highlight](https://microsoft.github.io/language-server-protocol/specification#textDocument_documentHighlight)
+      # informs the editor all relevant elements of the currently pointed item for highlighting. For example, when
+      # the cursor is on the `F` of the constant `FOO`, the editor should identify other occurrences of `FOO`
+      # and highlight them.
+      #
+      # For writable elements like constants or variables, their read/write occurrences should be highlighted
+      # differently. This is achieved by sending different "kind" attributes to the editor (2 for read and 3 for write).
+      #
+      # # Example
+      #
+      # ```ruby
+      # class Foo # when this is highlighted, the null highlight will be used to not match anything
+      # end
+      # ```
+      class NullHighlight < HighlightListener
+        extend T::Sig
 
-        match = T.must(@target).highlight_type(node)
-        add_highlight(match) if match
+        ResponseType = type_member { { fixed: T::Array[Interface::DocumentHighlight] } }
       end
 
-      private
+      class << self
+        extend T::Sig
 
-      sig { params(match: Support::HighlightTarget::HighlightMatch).void }
-      def add_highlight(match)
-        range = range_from_location(match.location)
-        @_response << Interface::DocumentHighlight.new(range: range, kind: match.type)
+        sig do
+          type_parameters(:ResponseType)
+            .params(
+              target: T.nilable(YARP::Node),
+              parent: T.nilable(YARP::Node),
+              emitter: EventEmitter,
+              message_queue: Thread::Queue,
+            ).returns(Listener[T::Array[Interface::DocumentHighlight]])
+        end
+        def for(target, parent, emitter, message_queue)
+          case target
```

Comment:
I'm a bit unsure about the document highlight refactor. Our current structure really doesn't account for having multiple listeners for the same request, as you probably notice when repeating the same documentation multiple times. And we still end up with a large case/when to select the right listener.

I think we can probably achieve a good refactor while keeping a single listener if we just focus on getting rid of `on_node`. I think it'd be fine to have the listener be instantiated with the target and then handle all of the different events where we look for highlights.